### PR TITLE
This adds thaw.js

### PR DESCRIPTION
**Pull request by freddyC**
A GIF to give the reviewer some spice of the internet

should solve https://github.com/BrainJS/brain.js/issues/82

### Removes `reinforce` option
I felt this cleans up and simplifies the network without changing functionality.

Basically when you train if it doesn't have a set size for the network it initializes and uses your options (as normal) to dictate network size, layers and all

If you have already trained the network or if the network is loaded from JSON then you don't need to set it and it trains as normal.

### Potential PROBLEMS
1) The train options `training callback called with training stats` test fails
    a) I think this is a problem in thaw.js but haven't been able to track it down yet.

2) Should probably have more tests for the new train options

3) With training being async now this may create a new problem with trying to train when currently training.  I haven't tested anything on that I was just thinking and about ramifications of this change.  We might want to add some sort of (is training) variable and if someone tries to train while it's currently training either push this training to the end of current training or escape out with an error or something.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -6,4 +6,6 @@ node_modules
 npm-debug.log
 
 # Yarn
-yarn.lock
\ No newline at end of file
+yarn.lock
+
+/dist**
```
Comment (freddyC) [inline .gitignore]: yeah I shouldn't have done this

**Comment by robertleeplummerjr** [comment]
Can we alter this to have have an async method, rather than changing the established method?  If we do this than the sync method stays as it is, we get a new feature, don't break backward compatibility.

**Comment by robertleeplummerjr** [comment]
Interesting idea on `reinforce`, so if the net is defined, `reinforce` is implied?

**Comment by freddyC** [comment]
Yeah, that would make the default behavior allow us to pick up and improve training at any point.

I think this will lead naturally to reinforcement especially if the user can set a required confidence level, and then things that we are less confident on the user can have it set up to trigger him and he lets the net know the correct answer and that can be used to reinforce.  

I will move the training to a train async method,  Are you ok with keeping the reinforcement as I set it up or do you want me to revert that change?

**Comment by robertleeplummerjr** [comment]
We should ask the community on the thread where `reinforce` was added, but I think they'll love it.

File: .gitignore (hunk 1 above)
Comment (robertleeplummerjr) [reply .gitignore]: Can you remove the dist

**Comment by freddyC** [comment]
So I made the changes but am working on the tests, and I think I broke stuff so I will try and get it updated this weekend, just wanted to keep momentum on this.

File: src/neural-network.js [hunk 2]
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
...
-      error: error,
-      iterations: i
+  /**
+   *
+   * @param data
+   * @param _options
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: Can we just use a `Promise` here?

File: src/neural-network.js [hunk 3]
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
...
+    let callback = options.callback;
+    let callbackPeriod = options.callbackPeriod;
+    let endTime = Date.now() + options.trainTimeMs;
+    let res = {
+      error: 1,
+      iterations: 0
     };
+
+    if (this.sizes === null) {
+      this.initialize (data);
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: The space after a method I've not seen in javascript libraries, can we aim for `justThis()`

File: src/neural-network.js [hunk 4]
```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
...
-      iterations: i
+  /**
+   *
+   * @param data
+   * @param _options
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
+    if (typeof _options === "function") {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: We've been in the practice of using single quotes, would you be ok migrating to such?

File: test/base/likely.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ import assert from 'assert';
 import brain from '../../src';
 import likely from '../../dist/likely';
 
-describe('likely', () => {
+describe('likely', function () {
```
Comment (robertleeplummerjr) [inline test/base/likely.js]: Lets keep the newer `() => {}` syntax where possible.

**Review by robertleeplummerjr — CHANGES_REQUESTED** [review]
Good job, had a few items to review.  Ty ty!

File: src/neural-network.js (hunk 4 above)
Comment (freddyC) [reply src/neural-network.js]: absolutely

File: test/base/likely.js (hunk 5 above)
Comment (freddyC) [reply test/base/likely.js]: One thing I found out is that mocha doesn't play well with the new `() => {` syntax.  I like it better but I was getting some really weird crashes that I think had to do with that. That being said I probably went overboard in changing it to `function () {`

File: src/neural-network.js [hunk 6]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+      cb = _options;
+      _options = {};
+    }
+    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+    data = this.formatData(data);
+    let iterations = options.iterations;
+    let errorThresh = options.errorThresh;
+    let log = options.log === true ? console.log : options.log;
+    let logPeriod = options.logPeriod;
+    let learningRate = _options.learningRate || this.learningRate || options.learningRate;
```
Comment (perkyguy) [inline src/neural-network.js]: Why the triple places this could exist?

File: src/neural-network.js [hunk 7]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+    const thaw  = new Thaw (items, {
+      delay: true,
+      each: () => {
+        res.iterations++;
+        let sum = 0;
+        for (var i = 0; i < data.length; ++i) {
+          sum += this.trainPattern (data[i].input, data[i].output, learningRate);
+        }
+
+        res.error = sum / data.length;
```
Comment (perkyguy) [inline src/neural-network.js]: Why not just `res.error = data.reduce(...)/data.length`?

File: src/neural-network.js [hunk 8]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+
+        if (log && (res.iterations % logPeriod === 0)) {
+          log(`iterations: ${res.iterations} training error: ${res.error}`);
+        }
+
+        if (callback && (res.iterations % callbackPeriod === 0)) {
+          // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+          callback(JSON.parse(JSON.stringify(res)));
```
Comment (perkyguy) [inline src/neural-network.js]: `Object.assign` would be a little clearer for this?

File: test/base/likely.js (hunk 5 above)
Comment (robertleeplummerjr) [reply test/base/likely.js]: Just use `--compilers js:babel-core/register` with mocha

File: test/base/bitwise.js [hunk 9]
```diff
@@ -5,46 +5,88 @@ let wiggle = 0.1;
...
-    console.log('output', output);
-    let target = data[i].output;
-    assert.ok(output < (target + wiggle) && output > (target - wiggle), 'failed to train ' + op + ' - output: ' + output + ' target: ' + target);
-  }
+  let res = net.train (data, { errorThresh: 0.003 });
+
+  data.forEach(d => {
+    var actual = net.run(d.input)
+    var expected = d.output;
+    assert.ok(actual < (expected + wiggle) && actual < (expected + wiggle), `failed to train "${op}" - expected: ${expected}, actual: ${actual}`);
```
Comment (perkyguy) [inline test/base/bitwise.js]: Copy-pasta error? You have the same thing tested in this &&

File: test/base/bitwise.js [hunk 10]
```diff
@@ -5,46 +5,88 @@ let wiggle = 0.1;
...
+  });
+}
+
+function testBitwiseAsync (data, op) {
+  let net = new brain.NeuralNetwork();
+  net.trainAsync (data, { errorThresh: 0.003 }, res => {
+    data.forEach(d => {
+      var actual = net.run(d.input)
+      var expected = d.output;
+      assert.ok(actual < (expected + wiggle) && actual < (expected + wiggle), `failed to train "${op}" - expected: ${expected}, actual: ${actual}`);
```
Comment (perkyguy) [inline test/base/bitwise.js]: Same thing tested?

File: test/base/bitwise.js (hunk 9 above)
Comment (perkyguy) [reply test/base/bitwise.js]: Why not just do a `.toFixed(1)`?

**Review by perkyguy — COMMENTED** [review]
What's the use case for the training time?

File: src/neural-network.js [hunk 11]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+
+      if (log && (res.iterations % logPeriod === 0)) {
+        log('iterations:', res.iterations, 'training error:', res.error);
       }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
+        // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+        callback (JSON.parse (JSON.stringify (res)));
```
Comment (perkyguy) [inline src/neural-network.js]: `Object.assign`?

File: src/neural-network.js [hunk 12]
```diff
@@ -223,13 +223,36 @@ export default class NeuralNetwork {
...
+   * @param data
+   * @returns sizes
+   */
+  _getSizesFromData (data) {
+    let sizes = [];
+    let inputSize = data[0].input.length;
+    let outputSize = data[0].output.length;
+    let hiddenSizes = this.hiddenSizes;
+    if (!hiddenSizes) {
+      sizes.push(Math.max(3, Math.floor(inputSize / 2)));
```
Comment (perkyguy) [inline src/neural-network.js:236]: If someone doesn't want hidden layers, the library will force it?

File: src/neural-network.js [hunk 13]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
-        log('iterations:', i, 'training error:', error);
+      res.error = sum / data.length;
+
+      if (log && (res.iterations % logPeriod === 0)) {
+        log('iterations:', res.iterations, 'training error:', res.error);
       }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
```
Comment (perkyguy) [inline src/neural-network.js]: The callback is for getting periodic training results? Not the finished trained model?

File: src/neural-network.js [hunk 14]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+          sum += this.trainPattern (data[i].input, data[i].output, learningRate);
+        }
+
+        res.error = sum / data.length;
+
+        if (log && (res.iterations % logPeriod === 0)) {
+          log(`iterations: ${res.iterations} training error: ${res.error}`);
+        }
+
+        if (callback && (res.iterations % callbackPeriod === 0)) {
```
Comment (perkyguy) [inline src/neural-network.js]: Like above, the callback is for partially trained nets?

File: src/neural-network.js (hunk 7 above)
Comment (freddyC) [reply src/neural-network.js]: I was talking to @robertleeplummerjr about it yesterday and we talked about the cost of closures.  They aren't crazy expensive but there is a cost to them and I don't usually take them into consideration.  I am open to either way but where it was simple to make the for loop and iterate I did that.

So I am open to either way :)

File: src/neural-network.js (hunk 8 above)
Comment (freddyC) [reply src/neural-network.js]: 👍  I didn't think of that but I will get it changed

File: src/neural-network.js [hunk 15]
```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
...
+
+    if (this.sizes === null) {
+      let sizes = this._getSizesFromData (data);
+      this.initialize (sizes);
+    }
+
+    const items = new Array(iterations);
+    const thaw  = new Thaw (items, {
+      delay: true,
+      each: () => {
```
Comment (perkyguy) [inline src/neural-network.js]: This is the same as the standard `train` behavior, just delivered per iteration. This should be pulled out into its own function/method to keep it DRY

File: test/base/bitwise.js (hunk 9 above)
Comment (freddyC) [reply test/base/bitwise.js]: Good idea on the `.toFixed (1)` and yeah that was meant to clamp that value.

File: src/neural-network.js (hunk 13 above)
Comment (freddyC) [reply src/neural-network.js]: yeah that's how it is currently setup.

File: src/neural-network.js (hunk 12 above)
Comment (freddyC) [reply src/neural-network.js:236]: yep, that's how it is currently setup.

File: src/neural-network.js (hunk 13 above)
Comment (perkyguy) [reply src/neural-network.js]: Why?

File: src/neural-network.js (hunk 15 above)
Comment (freddyC) [reply src/neural-network.js]: 👍

File: src/neural-network.js (hunk 12 above)
Comment (perkyguy) [reply src/neural-network.js:236]: Why? Can @robertleeplummerjr she's some light on this?

File: src/neural-network.js (hunk 6 above)
Comment (freddyC) [reply src/neural-network.js]: The only thing I can think of is that it gives preference to as so:
1) the passed in learning rate
2) the class's set learning rate
3) the default learning rate.

It is a little odd to me but I figured it was outside the scope of this pr

File: src/neural-network.js (hunk 13 above)
Comment (robertleeplummerjr) [reply src/neural-network.js]: logging for the most part.  Most people like to see training advance.  I don't see this going away any time soon, as it was part of the model when the community took over.

File: src/neural-network.js (hunk 11 above)
Comment (robertleeplummerjr) [reply src/neural-network.js]: or `{...res}`, which is essentially the same thing.

File: src/neural-network.js (hunk 15 above)
Comment (robertleeplummerjr) [reply src/neural-network.js]: I agree, that'd be more ideal.

File: src/neural-network.js (hunk 14 above)
Comment (robertleeplummerjr) [reply src/neural-network.js]: that is correct as well

**Comment by robertleeplummerjr** [comment]
all on all I see some really cool things happening here.

File: src/neural-network.js (hunk 12 above)
Comment (perkyguy) [reply src/neural-network.js:236]: Sorry, I read this as "yeah I did it this way, what of it?" Not as you probably intended it, "maintaining current functionality"

File: src/neural-network.js (hunk 13 above)
Comment (perkyguy) [reply src/neural-network.js]: Isn't that what the `log` option does? It's cool though, separate conversation for the community, this maintains current functionality

File: src/neural-network.js (hunk 12 above)
Comment (freddyC) [reply src/neural-network.js:236]: haha no worries, it's finding the balance between not changing too much but still changing things in a good way :)

**Comment by freddyC** [comment]
~@robertleeplummerjr @perkyguy this I think is ready.  The only thing I am not real happy with is the log tests.  They pass but I don't know how to get them to log out correct.~

image

~this is what they look like right now.  Mocha is logging something out and I don't know exactly what it's doing.  Ideally I think we should be allowing the options to contain a function for log and call their function instead of just take a true or false and then using console.log.  That would allow someone to plug in their custom logger.  That I figured was outside the scope of this pr lol.~

**Comment by freddyC** [comment]
I actually just added in the change for logging, it felt right, hopefully it is ok (a bit of scope creep but worse things have happened, I guarantee it).

File: src/neural-network.js [hunk 16]
```diff
@@ -63,7 +63,7 @@ export default class NeuralNetwork {
    *
    * @param {Number[]} sizes
    */
-  initialize(sizes) {
+  initialize (sizes) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 17]
```diff
@@ -223,13 +223,36 @@ export default class NeuralNetwork {
...
+  }
+
   /**
    *
    * @param data
    * @param _options
    * @returns {{error: number, iterations: number}}
    */
-  train(data, _options = {}) {
+  train (data, _options = {}) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 18]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
     return output;
   }
 
+  /**
+   *
+   * @param data
+   * @returns sizes
+   */
+  _getSizesFromData (data) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: uneeded space

File: src/neural-network.js (hunk 12 above)
Comment (robertleeplummerjr) [reply src/neural-network.js:236]: Thats how it was when we got here, however one could do: `{ hiddenSizes: [] }`

File: src/neural-network.js [hunk 19]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+
+  /**
+   *
+   * @param log
+   * if true passed in console.log is used
+   * if a method is passed in method is used
+   * if false passed in nothing is logged
+   * @returns error
+   */
+  _setLogMethod (log) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 20]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+    return false;
+  }
+
+  /**
+   *
+   * @param data
+   * @param learning Rate
+   * @returns error
+   */
+  _calculateTrainingError (data, learningRate) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 21]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+    let sum = 0;
+    for (let i = 0; i < data.length; ++i) {
+      sum += this.trainPattern (data[i].input, data[i].output, learningRate);
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 22]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+    }
+    return sum / data.length;
+  }
+
+  /**
+   *
+   * @param status { iterations: number, error: number}
+   * @param options
+   */
+  _checkTrainingTick (data, status, options) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 23]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+
   /**
    *
    * @param data
    * @param _options
    * @returns {{error: number, iterations: number}}
    */
-  train(data, _options = {}) {
-    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+  train (data, _options = {}) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 24]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+    let options = Object.assign({}, this.constructor.trainDefaults, _options);
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: `options` is never reassigned, use a `const`, not a `let`.

File: src/neural-network.js [hunk 25]
```diff
@@ -22,7 +22,7 @@ export default class NeuralNetwork {
       learningRate: 0.3,
       callback: null,
       callbackPeriod: 10,
-      reinforce: false
+      trainTimeMs: -Infinity
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: Can the `trainTimeMs` be put into a seperate pr?  Want to keep the pr's as lean as possible.

File: src/neural-network.js [hunk 26]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
-        let err = this.trainPattern(data[j].input, data[j].output, learningRate);
-        sum += err;
+  /**
+   *
+   * @param data
+   * @param _options
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space, this line and next

File: src/neural-network.js [hunk 27]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+      options.learningRate = _options.learningRate || this.learningRate || options.learningRate;
+      let endTime = Date.now() + options.trainTimeMs;
+
+      let status = {
+        error: 1,
+        iterations: 0
+      };
+
+      if (this.sizes === null) {
+        let sizes = this._getSizesFromData (data);
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space, this line and next.

File: src/neural-network.js [hunk 28]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
 
-      if (log && (i % logPeriod === 0)) {
-        log('iterations:', i, 'training error:', error);
-      }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
-      }
-    }
+      const items = new Array(options.iterations);
+      const thaw  = new Thaw (items, {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 29]
```diff
@@ -223,63 +223,145 @@ export default class NeuralNetwork {
...
+        delay: true,
+        each: () => {
+          this._checkTrainingTick (data, status, options);
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 30]
```diff
@@ -288,7 +370,7 @@ export default class NeuralNetwork {
    * @param target
    * @param learningRate
    */
-  trainPattern(input, target, learningRate) {
+  trainPattern (input, target, learningRate) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js]: unneeded space

File: src/neural-network.js [hunk 31]
```diff
@@ -634,8 +716,8 @@ export default class NeuralNetwork {
    * @param json
    * @returns {NeuralNetwork}
    */
-  fromJSON(json) {
-    this.initialize(json.sizes);
+  fromJSON (json) {
```
Comment (robertleeplummerjr) [inline src/neural-network.js:717]: unneeded space

File: test/base/bitwise.js [hunk 32]
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
 
 let wiggle = 0.1;
 
+function isAround (actual, expected) {
```
Comment (robertleeplummerjr) [inline test/base/bitwise.js]: unneeded space

File: test/base/bitwise.js [hunk 33]
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
...
+
+  data.forEach(d => {
+    var actual = net.run(d.input)
+    var expected = d.output;
+    assert.ok(isAround (actual, expected), `failed to train "${op}" - expected: ${expected}, actual: ${actual}`);
+  });
 }
 
-describe('bitwise functions', () => {
+function testBitwiseAsync (data, op, done) {
```
Comment (robertleeplummerjr) [inline test/base/bitwise.js]: unneeded space

File: test/base/bitwise.js [hunk 34]
```diff
@@ -3,48 +3,102 @@ import brain from '../../src';
...
+  let net = new brain.NeuralNetwork();
+  net
+    .trainAsync (data, { errorThresh: 0.003 })
```
Comment (robertleeplummerjr) [inline test/base/bitwise.js]: there are several more of these, can you just handle all of them?

File: src/neural-network.js (hunk 31 above)
Comment (robertleeplummerjr) [reply src/neural-network.js:717]: still one space 😄

File: test/base/hash.js [hunk 35]
```diff
@@ -2,79 +2,174 @@ import assert from 'assert';
 import brain from '../../src';
 
 describe('hash input and output', () => {
-  it('runs correctly with array input and output', () => {
+  it ('runs correctly with array input and output', () => {
```
Comment (robertleeplummerjr) [inline test/base/hash.js:5]: (space)

File: test/base/likely.js [hunk 36]
```diff
@@ -62,23 +66,24 @@ describe('likely', () => {
...
-    let result = likely(character(
-      '######.' +
-      '#.....#' +
-      '#.....#' +
-      '######.' +
-      '#..#..#' +
-      '#.....#' +
-      '###.##.'
-    ), net);
+      let result = likely(character(
```
Comment (robertleeplummerjr) [inline test/base/likely.js:69]: spacing is odd here

**Review by robertleeplummerjr — APPROVED** [review]
This looks great.  I had a couple suggestions, but they don't change functionality, it was just spacing.  Ty for your hard work!

**Comment by robertleeplummerjr** [comment]
ty so much for your hard work on this!  Sorry about the conflicts, I went ahead and did a merge, tested, and merged.

**Comment by freddyC** [comment]
giphy

**Pull request merged**